power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_se_solver.hpp

```diff
@@ -163,38 +178,506 @@ template <bool sym> class NewtonRaphsonSESolver {
+        block.r_Q_v() = RealTensor<sym>{RealValue<sym>{-injection.q_variance}};
+    }
+
+    /// @brief Processes common part of all elements to fill from an injection measurement.
+    ///
+    /// Also includes zero injection constraint.
+    /// This would be H, N, M, L at (row, col) block and partially the second part of the (row, row) block using the
+    /// same H and M but multiplied by abs_ui_inv.
+    ///
+    /// @param block LHS(r, c)
+    /// @param diag_block LHS(r, r)
+    /// @param rhs_block RHS(r)
+    /// @param yij admittance of (row with injection, c)
+    /// @param u_state Voltage state of iteration
+    void process_injection_row(NRSEGainBlock<sym>& block, NRSEGainBlock<sym>& diag_block, NRSERhs<sym>& rhs_block,
+                               auto const& yij, auto const& u_state) {
+        auto const hm_ui_uj_yij = hm_complex_form(yij, u_state.ui_uj_conj);
+        auto const nl_ui_uj_yij = dot(hm_ui_uj_yij, u_state.abs_uj_inv);
+        auto const injection_jac = calculate_jacobian(hm_ui_uj_yij, nl_ui_uj_yij);
+        add_injection_jacobian(block, injection_jac);
+
+        // add paritial sum to the diagonal block and subtract from rhs for current row
+        auto const f_x_complex_row = sum_row(hm_ui_uj_yij);
+        auto const f_x_complex_abs_ui_inv_row = dot(u_state.abs_ui_inv, f_x_complex_row);
+        auto const injection_jac_diagonal = jacobian_diagonal_component(f_x_complex_abs_ui_inv_row, f_x_complex_row);
+        add_injection_jacobian(diag_block, injection_jac_diagonal);
+        rhs_block.tau_p() -= real(f_x_complex_row);
+        rhs_block.tau_q() -= imag(f_x_complex_row);
+    }
+
+    /// @brief Adds contribution of G(i, i) form the shunt.
+    /// Note: The sign of Y_s is inverted per injection direction.
+    ///
+    /// @param block LHS(i, i)
+    /// @param rhs_block RHS(i)
+    /// @param ys shunt admittance related to the shunt measurement
+    /// @param u_state Voltage state of iteration Voltage state of iteration
+    /// @param measured_power measured shunt power
+    void process_shunt_measurement(NRSEGainBlock<sym>& block, NRSERhs<sym>& rhs_block, auto const& ys,
+                                   auto const& u_state, auto const& measured_power) {
+        auto const hm_ui_ui_ys = hm_complex_form(-ys, u_state.ui_ui_conj);
+        auto const nl_ui_ui_ys = dot(hm_ui_ui_ys, u_state.abs_ui_inv);
+        auto const f_x_complex = sum_row(hm_ui_ui_ys);
+        auto const f_x_complex_abs_ui_inv = sum_row(nl_ui_ui_ys);
+
+        auto jac_block = calculate_jacobian(hm_ui_ui_ys, nl_ui_ui_ys);
+        jac_block += jacobian_diagonal_component(f_x_complex_abs_ui_inv, f_x_complex);
+        auto const& block_F_T_k_w = transpose_multiply_weight(jac_block, measured_power);
+        multiply_add_jacobian_blocks_lhs(block, block_F_T_k_w, jac_block);
+        multiply_add_jacobian_blocks_rhs(rhs_block, block_F_T_k_w, measured_power, f_x_complex);
+    }
```

Comment:
I think the `i` suffix added to variable names can be dropped:
 - the context is clear that it is for diagonal blocks
 - there is no `j` to differrentiate from

Reply:
i think for consistency keeping it as is is OK. we can always change this in the future.

Reply:
Agreed. Let's keep it for later discussion.